src/rgw/rgw_putobj_processor.cc

```diff
@@ -497,13 +497,24 @@ int MultipartObjectProcessor::complete(size_t accounted_size,
     return r;
   }
 
-  encode(info, bl);
-
   std::unique_ptr<rgw::sal::Object> meta_obj =
     head_obj->get_bucket()->get_object(rgw_obj_key(mp.get_meta(), std::string(), RGW_OBJ_NS_MULTIPART));
   meta_obj->set_in_extra_data(true);
 
-  r = meta_obj->omap_set_val_by_key(dpp, p, bl, true, null_yield);
+  rgw_raw_obj meta_raw_obj;
+  store->getRados()->obj_to_raw(meta_obj->get_bucket()->get_placement_rule(), 
+                                meta_obj->get_obj(),
+                                &meta_raw_obj);
+  rgw_rados_ref meta_obj_ref;
+  r = store->getRados()->get_raw_obj_ref(dpp, meta_raw_obj, &meta_obj_ref);
+  if (r < 0) {
+    ldpp_dout(dpp, -1) << "ERROR: failed to get obj ref of meta obj with ret=" << r << dendl;
+    return r;
+  }
+
+  r = cls_rgw_mp_upload_part_info_update(meta_obj_ref.pool.ioctx(), meta_obj_ref.obj.oid, p, info);
```

Comment:
this could be problematic during rolling upgrades. if the osd hasn't yet upgraded to a version that understands this new `rgw_mp_upload_part_info_update` call, we'd expect to get a [-EOPNOTSUPP](https://github.com/ceph/ceph/blob/cb17f286/src/osd/PrimaryLogPG.cc#L6160-L6163) error in response. in that case, we should fall back to the original `omap_set_val_by_key()` call

Reply:
maybe it should just return EIO